src/main/scala/org/fortysevendeg/sparksftp/ReadingSFTPConnectorApp.scala

```diff
@@ -18,41 +17,70 @@ object ReadingSFTPConnectorApp extends IOApp {
     ConfigLoader[IO]
       .loadConfig[ReadingSFTPConfig]
 
-  def run(args: List[String]): IO[ExitCode] =
-    for {
-      config: ReadingSFTPConfig <- setupConfig
-      defaultSparkConf: SparkConf = SparkUtils.createSparkConfWithSFTPSupport(config)
-      sparkSession: SparkSession = SparkSession.builder
+  def run(args: List[String]): IO[ExitCode] = {
+
+    def createSparkSession(config: ReadingSFTPConfig): IO[SparkSession] = IO {
+      val defaultSparkConf: SparkConf = createSparkConfWithSFTPSupport(config)
+      SparkSession.builder
         .config(defaultSparkConf)
         .enableHiveSupport
         .getOrCreate()
+    }
 
-      sftpConfig = configs.SFTPConfig
-        .configFromContextProperties(sparkSession.sparkContext, config.sftp)
-
-      // Read the source files from SFTP into dataframes
-      users = dataframeFromCsvWithSFTPConnector(sparkSession, sftpConfig, sftpConfig.sftpUserPath)
-      salaries = dataframeFromCsvWithSFTPConnector(
-        sparkSession,
-        sftpConfig,
-        sftpConfig.sftpSalaryPath
-      )
+    def readDataFramesWithSFTPConnector(
+        sparkSession: SparkSession,
+        sftpConfig: SFTPConfig
+    ): IO[(DataFrame, DataFrame)] =
+      for {
+        users <- dataframeFromCsvWithSFTPConnector(
+          sparkSession,
+          sftpConfig,
+          sftpConfig.sftpUserPath
+        )
+        salaries <- dataframeFromCsvWithSFTPConnector(
+          sparkSession,
+          sftpConfig,
+          sftpConfig.sftpSalaryPath
+        )
+      } yield (users, salaries)
 
-      // Sample operations to persist and query the Hive database
-      _                                        = HiveUserData.persistUserData(sparkSession, users, salaries)
-      (userDataFromHive, salariesDataFromHive) = HiveUserData.readUserData(sparkSession)
+    def toSFTPCompressedCSVWithSFTPConnector(
+        userData: DataFrame,
+        salariesData: DataFrame,
+        userNewSalaries: DataFrame,
+        sftpConfig: SFTPConfig
+    ): IO[Unit] =
+      for {
+        _ <- dataframeToCompressedCsvWithSFTPConnector(
+          userData,
+          sftpConfig,
+          s"${sftpConfig.sftpUserPath}_output"
+        )
+        _ <- dataframeToCompressedCsvWithSFTPConnector(
+          salariesData,
+          sftpConfig,
+          s"${sftpConfig.sftpSalaryPath}_output"
+        )
+        _ <- dataframeToCompressedCsvWithSFTPConnector(
+          userNewSalaries,
+          sftpConfig,
+          s"${sftpConfig.sftpSalaryPath}_transformed_output"
+        )
+      } yield ()
```

Comment:
These 3 operations are running sequentially, does it make sense if we run them in parallel?

```scala
import cats.syntax.parallel._

// rest of the code

(
        dataframeToCompressedCsvWithSFTPConnector(
          userData,
          sftpConfig,
          s"${sftpConfig.sftpUserPath}_output"
        ),
        dataframeToCompressedCsvWithSFTPConnector(
          salariesData,
          sftpConfig,
          s"${sftpConfig.sftpSalaryPath}_output"
        ),
        dataframeToCompressedCsvWithSFTPConnector(
          userNewSalaries,
          sftpConfig,
          s"${sftpConfig.sftpSalaryPath}_transformed_output"
        )
      ).parMapN((_, _, _) => ())
```

Reply:
Very good question.

That was my first approach, that was I still had a ContextShift implicit, needed for the parMapN operations, until I thought about the problems I had when exporting to the same SFTP folder destination, basically the same issue that raised that led me to do the coalescing. 
We could do that approach, could we expand that option in another branch, with its respective issue, testing it, specially with larger files, that is when partitions come into play, takes a lot of time.
The purpose of this branch was to have several executors on place specially to test paralellism for the intermediate stages, not the ingestion or the exporting.

What do you think?